{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee27cddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Computing model baselines on real diabetes data\n",
    "# Using full size dataset (4901 samples x 72819 SNPs)\n",
    "\n",
    "# Velina Kozareva"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7fa0512c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import joblib\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.model_selection import cross_validate\n",
    "\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "from sklearn import metrics\n",
    "\n",
    "import matplotlib.pyplot as plt \n",
    "\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import roc_curve"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43a988e8",
   "metadata": {},
   "source": [
    "## Modifications to Divya's model baseline code\n",
    "\n",
    "Once this is organized in repo, can make the corresponding changes there and redo notebook more cleanly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "381ea0f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ModelSum:\n",
    "    \"\"\"Handy class to store all information about each model\"\"\"\n",
    "    def __init__(self, pset, simnum):\n",
    "        self.pset = pset\n",
    "        self.simnum = int(simnum)\n",
    "\n",
    "        self.data_type = None\n",
    "        # self.log_reg_real = None\n",
    "        # self.rand_forest_real = None\n",
    "        # self.log_reg_rand = None\n",
    "        # self.rand_forest_rand = None\n",
    "        self.rho=None\n",
    "        self.pve=None\n",
    "        self.maf_frac=None\n",
    "        self.obs=None\n",
    "        self.k=None\n",
    "        self.overlap=None\n",
    "        self.class_imbalance = 0.5\n",
    "\n",
    "        self.real_interp_auc = {}\n",
    "        self.rand_interp_auc = {}\n",
    "\n",
    "        self.real_class_auc = {}\n",
    "        self.rand_class_auc = {}\n",
    "\n",
    "        self.classification_aucs = {}\n",
    "        ### Return importance scores as well\n",
    "        self.vscores = {}\n",
    "\n",
    "    def populate_pset_vars(self):\n",
    "        # Set up simlation run\n",
    "        sp = self.pset.split('-')\n",
    "\n",
    "        if 'prop_case' in sp:\n",
    "            self.class_imbalance = float(sp[sp.index('prop_case')+1])\n",
    "        if 'rho' in sp:\n",
    "            self.rho = float(sp[sp.index('rho')+1])\n",
    "        if 'pve' in sp:\n",
    "            self.pve = float(sp[sp.index('pve')+1])\n",
    "        if 'maf_frac' in sp:\n",
    "            self.maf_frac = float(sp[sp.index('maf_frac')+1])\n",
    "        if 'obs' in sp:\n",
    "            self.obs = float(sp[sp.index('obs')+1])\n",
    "        if 'k' in sp:\n",
    "            self.k = float(sp[sp.index('k')+1])\n",
    "\n",
    "        if 'ind' in sp:\n",
    "            self.overlap = True\n",
    "        elif 'non_overlap_degree' in sp:\n",
    "            self.overlap = False\n",
    "\n",
    "# get interpretability and class ROC for one split (then average ROCs over 5 splits)\n",
    "\n",
    "def split_data(X, y, test_size=0.1, rand_state=5):\n",
    "    train_x, test_x, train_y, test_y = train_test_split(X, y, test_size=test_size, random_state=rand_state,\n",
    "                                                    shuffle = True, stratify = y)\n",
    "    snp_labels = list(X.columns)\n",
    "    return train_x, test_x, train_y, test_y, snp_labels\n",
    "\n",
    "\n",
    "def train_model(mod_type, train_x, test_x, train_y, test_y, var_labels=None, causal_vars=None, var_type='snps', rand_state=5, calc_interpretability=True):\n",
    "    # Setup classifier\n",
    "    implemented_list = ['log_reg', 'rand_forest']\n",
    "#     if not calc_interpretability:\n",
    "    interp_auc = None\n",
    "\n",
    "    assert mod_type in implemented_list, 'check mod type: {}'.format(mod_type)\n",
    "    if mod_type == 'log_reg':\n",
    "        mlmod = LogisticRegression(penalty='l2', solver='liblinear', class_weight='balanced', random_state=rand_state)\n",
    "        mlmod.fit(train_x, np.ravel(train_y)) \n",
    "        test_class_auc = mlmod.score(test_x, test_y) #confirm that score is AUC\n",
    "\n",
    "        if calc_interpretability:\n",
    "            if var_type == 'snps':\n",
    "                coefs = np.abs(mlmod.coef_) # check absolute value\n",
    "#           elif causa_vars == 'pathways':\n",
    "            vscores = pd.DataFrame(zip(var_labels, np.ravel(coefs)), columns=['var', 'score']).set_index('var')\n",
    "            if causal_vars is not None:\n",
    "                _, interp_auc = calc_ROC(vscores, causal_vars)\n",
    "            \n",
    "    elif mod_type == 'rand_forest':\n",
    "        mlmod = RandomForestClassifier(class_weight='balanced', random_state=rand_state)\n",
    "        mlmod.fit(train_x, np.ravel(train_y)) \n",
    "        test_class_auc = mlmod.score(test_x, test_y) #confirm that score is AUC\n",
    "\n",
    "        if calc_interpretability:\n",
    "            if var_type == 'snps':\n",
    "                coefs = np.abs(mlmod.feature_importances_) # check absolute value\n",
    "                vscores = pd.DataFrame(zip(var_labels, np.ravel(coefs)), columns=['var', 'score']).set_index('var')\n",
    "            # elif var_type == 'patways':\n",
    "            if causal_vars is not None:\n",
    "                _, interp_auc = calc_ROC(vscores, causal_vars)\n",
    "        \n",
    "    ### Return vscores as well\n",
    "    return mlmod, test_class_auc, interp_auc, vscores\n",
    "\n",
    "# right now, using snps; modify so can use either SNps or genens\n",
    "def calc_ROC(df_scores, true_causal): # layer: 'snp' or 'gene', list_pos: 'list_snp' or 'list_genes'\n",
    "    # from Ashley\n",
    "    assert true_causal is not None, 'check true_causal list'\n",
    "\n",
    "    npips = len(df_scores)\n",
    "    all_vals = list(df_scores.index)\n",
    "    Pos = true_causal\n",
    "    Negs = set(all_vals) - set(Pos)\n",
    "\n",
    "    df_res = pd.DataFrame(columns=[\"TPR\", \"FPR\", \"FDR\", \"PWR\"], dtype=object) \n",
    "    df_scores.sort_values(by=['score'], ascending=False, inplace=True)\n",
    "    \n",
    "    for i in range(1,npips+1):\n",
    "        v = df_scores[0:i].index\n",
    "        z = set(all_vals) -  set(v)\n",
    "#         TP = len(intersection(set(v), set(Pos)))\n",
    "        TP = len(set(v) & set(Pos))\n",
    "        FP = len(set(v)-set(Pos))\n",
    "#         TN = len(intersection(set(z), set(Negs)))\n",
    "        TN = len(set(z) & set(Negs))\n",
    "        FN = len(set(z)-set(Negs))\n",
    "#         print(TPs,FP,TN,FN)        \n",
    "        TPR = TP/(TP+FN)\n",
    "            \n",
    "        FPR = FP/(FP+TN) \n",
    "        FDR = FP/(FP+TP)\n",
    "        PWR = 1-(1-TPR)\n",
    "        df_res.loc[len(df_res)] = [TPR, FPR, FDR, PWR]\n",
    "        \n",
    "    # add AUC calc here    \n",
    "    auc = metrics.auc(np.array(df_res['FPR']), np.array(df_res['TPR']))\n",
    "    return df_res, auc\n",
    "\n",
    "def score_cv(model_type, X, y, true_causal=None, calc_interpretability=False, cv=5):\n",
    "    mods = []\n",
    "    class_aucs = []\n",
    "    interp_aucs = []\n",
    "    ### Keep track of vscores as well\n",
    "    vscores_l = []\n",
    "    \n",
    "    for i in range(cv):\n",
    "        train_x, test_x, train_y, test_y, snp_labels = split_data(X, y, rand_state=5+i) #check random state set\n",
    "        mlmod, test_class_auc, interp_auc, vscores = train_model(model_type, train_x, test_x, train_y, test_y, snp_labels, true_causal, calc_interpretability=calc_interpretability) \n",
    "\n",
    "        # print(i, test_class_auc, interp_auc)\n",
    "        mods.append(mlmod)\n",
    "        class_aucs.append(test_class_auc)\n",
    "        interp_aucs.append(interp_auc)\n",
    "        vscores_l.append(vscores)\n",
    "        \n",
    "    return mods, class_aucs, interp_aucs, vscores_l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c85714b2",
   "metadata": {},
   "source": [
    "## Result-generating code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6616271",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in diabetes data\n",
    "\n",
    "data_pd = joblib.load('/scratch/users/velina/sparse_nn/t1d_data/pathway_level_data_snps72819/data_pd.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "93e7c1c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rs6658795</th>\n",
       "      <th>rs1016209</th>\n",
       "      <th>rs760967</th>\n",
       "      <th>rs760969</th>\n",
       "      <th>rs2235550</th>\n",
       "      <th>rs6686862</th>\n",
       "      <th>rs6687289</th>\n",
       "      <th>rs6694932</th>\n",
       "      <th>rs7534396</th>\n",
       "      <th>rs7542174</th>\n",
       "      <th>...</th>\n",
       "      <th>rs3865507</th>\n",
       "      <th>rs17771961</th>\n",
       "      <th>rs3928916</th>\n",
       "      <th>rs3928917</th>\n",
       "      <th>rs7251886</th>\n",
       "      <th>rs10423171</th>\n",
       "      <th>rs11883018</th>\n",
       "      <th>rs2701</th>\n",
       "      <th>rs3795018</th>\n",
       "      <th>rs11669653</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ind0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ind1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ind2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ind3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ind4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 72819 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      rs6658795  rs1016209  rs760967  rs760969  rs2235550  rs6686862  \\\n",
       "ind0        0.0        0.0       2.0       2.0        0.0        0.0   \n",
       "ind1        0.0        0.0       1.0       1.0        0.0        0.0   \n",
       "ind2        0.0        1.0       1.0       1.0        1.0        0.0   \n",
       "ind3        0.0        1.0       0.0       0.0        1.0        0.0   \n",
       "ind4        0.0        0.0       0.0       0.0        0.0        0.0   \n",
       "\n",
       "      rs6687289  rs6694932  rs7534396  rs7542174  ...  rs3865507  rs17771961  \\\n",
       "ind0        0.0        0.0        0.0        0.0  ...        0.0         0.0   \n",
       "ind1        0.0        0.0        0.0        0.0  ...        0.0         0.0   \n",
       "ind2        0.0        0.0        1.0        1.0  ...        0.0         0.0   \n",
       "ind3        0.0        0.0        1.0        1.0  ...        0.0         0.0   \n",
       "ind4        0.0        0.0        0.0        0.0  ...        0.0         0.0   \n",
       "\n",
       "      rs3928916  rs3928917  rs7251886  rs10423171  rs11883018  rs2701  \\\n",
       "ind0        1.0        1.0        2.0         0.0         0.0     1.0   \n",
       "ind1        2.0        1.0        1.0         0.0         0.0     1.0   \n",
       "ind2        2.0        2.0        2.0         0.0         0.0     1.0   \n",
       "ind3        2.0        1.0        2.0         0.0         0.0     1.0   \n",
       "ind4        1.0        0.0        0.0         0.0         0.0     0.0   \n",
       "\n",
       "      rs3795018  rs11669653  \n",
       "ind0        0.0         2.0  \n",
       "ind1        0.0         2.0  \n",
       "ind2        1.0         1.0  \n",
       "ind3        0.0         2.0  \n",
       "ind4        1.0         1.0  \n",
       "\n",
       "[5 rows x 72819 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ind0</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ind1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ind2</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ind3</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ind4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      label1\n",
       "ind0       0\n",
       "ind1       0\n",
       "ind2       0\n",
       "ind3       0\n",
       "ind4       0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Reformat data\n",
    "X = data_pd.drop(['label1', 'label2'], axis=1)\n",
    "display(X.head())\n",
    "y = data_pd[['label1']]\n",
    "display(y.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f236e214",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 51min 41s, sys: 34min 19s, total: 1h 26min\n",
      "Wall time: 45min 35s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Run data through baseline models\n",
    "ms = ModelSum('diabetes', 1)\n",
    "ms.populate_pset_vars()\n",
    "\n",
    "mod_list = ['log_reg', 'rand_forest']\n",
    "cX = X\n",
    "cY = y\n",
    "\n",
    "for model in mod_list:\n",
    "    mmod, mcauc, miauc, vscores = score_cv(model, cX, cY, true_causal=None, calc_interpretability=True)\n",
    "    ms.real_class_auc[model]=mcauc\n",
    "    ms.real_interp_auc[model]=miauc\n",
    "    ms.vscores[model] = vscores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "42722302",
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir diabetes_baselines/rand_forest\n",
    "! mkdir diabetes_baselines/log_reg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "25ca2e49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save these with appropriate names for next step (GSEA in R)\n",
    "for model in mod_list:\n",
    "    prefix = 'rf' if model == 'rand_forest' else 'lr'\n",
    "    for seed in range(5):\n",
    "        # Rename index for R function\n",
    "        ms.vscores[model][seed].index.names = ['X']\n",
    "        ms.vscores[model][seed].columns = ['coefs']\n",
    "        ms.vscores[model][seed].to_csv(f'diabetes_baselines/{model}/{prefix}_s{seed}_coefs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d5602667",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'log_reg': [0.6191446028513238,\n",
       "  0.6232179226069247,\n",
       "  0.6272912423625254,\n",
       "  0.6130346232179226,\n",
       "  0.6374745417515275],\n",
       " 'rand_forest': [0.604887983706721,\n",
       "  0.6089613034623218,\n",
       "  0.615071283095723,\n",
       "  0.6089613034623218,\n",
       "  0.6171079429735234]}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the classification AUCs here?\n",
    "# Quite a bit lower than expected...\n",
    "ms.real_class_auc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "843c7e18",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (als-genomics)",
   "language": "python",
   "name": "als-genomics"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
